### PR TITLE
Multiclass Logistic Regression Model

### PR DESCRIPTION
**Pull request by Sar-thak-3**

### CODE REVIEW

**Review by harshjohar — CHANGES_REQUESTED** [review]
Do the following steps
- pull the latest commit from `main` branch
- resolve any conflicts
- run the script as mentioned the `README.md` point 9 to format the spacing and everything properly

**Comment by Sar-thak-3** [comment]
> * run the script as mentioned the `README.md` point 9 to format the spacing and everything properly
Sir should i Run "./clang-format-all src/" this command directly in my vs code terminal?

**Comment by uttammittal02** [comment]
> > * run the script as mentioned the `README.md` point 9 to format the spacing and everything properly
> >   Sir should i Run "./clang-format-all src/" this command directly in my vs code terminal?

Yes run it in the directory of your cloned repo

**Comment by harshjohar** [comment]
> > * run the script as mentioned the `README.md` point 9 to format the spacing and everything properly
> 
> Sir should i Run "./clang-format-all src/" this command directly in my vs code terminal?
> 
> 

If you have mac or linux then directly in terminal, but if you have windows, you gotta run this in git bash in the project root directory

**Comment by Sar-thak-3** [comment]
> > > * run the script as mentioned the `README.md` point 9 to format the spacing and everything properly
> > 
> > 
> > Sir should i Run "./clang-format-all src/" this command directly in my vs code terminal?
> 
> If you have mac or linux then directly in terminal, but if you have windows, you gotta run this in git bash in the project root directory

Screenshot 2023-02-08 125736

**Review by harshjohar — APPROVED** [review]
lgtm

**Review by ken1000minus7 — APPROVED** [review]
Lgtm gw

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Doesn't seem upto mark

**Comment by Sar-thak-3** [comment]
> Doesn't seem upto mark

Kindly review the changes

**Comment by uttammittal02** [comment]
> > Doesn't seem upto mark
> 
> Kindly review the changes

Bro the code isn't readable...aake bta ke ja ki kya kr rha....variable names thode readable rakh....Add some comments too

**Review by harshjohar — APPROVED** [review]
Lgtm

**Comment by uttammittal02** [comment]
@Sar-thak-3 Refer these [docs](https://github.com/PEC-CSS/slowmokit/blob/main/docs/methods/linear_model/linear_regression.md) for reference and add similar docs for your code

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+std::vector<double> LogisticRegression<T>::softmax(std::vector<T> x)
+{ // Softmax function
+  std::vector<double> exps(x.size());
+  double sum = 0.0;
+  for (int i = 0; i < x.size(); i++)
+  { // Iterating over vector to compute sum of all exponential values in vector
+    exps[i] = exp(x[i]);
+    sum += exps[i];
+  }
+  std::vector<double> values(x.size());
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:19]: Remove this vector declaration

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+};
+
+template<class T>
+double LogisticRegression<T>::crossEntropy(std::vector<int> y,
+                                           std::vector<T> qi)
+{ // Gives gradient descent -(d(z)/d(theta))
+  double l = 0.0;
+  for (int i = 0; i < y.size(); i++)
+  {
+    l += (y[i] * qi[i]);
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:34]: It should be `y[i] * log(q[i])` here. Read about logloss function used for logistic regression

File: docs/methods/linear_model/logistic_regression.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# Logistic Regression
+
+Logistic Regression model using stocastic gradient descent.
```
Comment (Ishwarendra) [inline docs/methods/linear_model/logistic_regression.md:3]: Typo: `stochastic` instead of `stocastic`

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    loss += crossEntropy(yi, qi); // Computing loss using z[i](computed values)
+                                  // and y[i](actual values)
+  }
+  return loss; // finally returning loss
+};
+
+template<class T>
+// This function performs logistic regression stochastic gradient descent
+std::vector<std::vector<double>>
+LogisticRegression<T>::logRegSgd(std::vector<std::vector<double>> x,
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:81]: Take user parameter `batch_size` and do the training in batches

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+};
+
+template<class T>
+// training function
+// x -> training 2-d vector double values
+// y - > training 1-d int values
+// alpha -> double which is going to be multiply with gradient descent
+// numEpochs -> number of epochs
+// verbose -> whether to show which epoch is going on
+void LogisticRegression<T>::train(std::vector<std::vector<T>> x,
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:175]: We'll be assuming that user won't provide onehot-encoded labels. You need to do that inside your code

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Good job....just a few more changes

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 6]
```diff
@@ -31,7 +30,7 @@ double LogisticRegression<T>::crossEntropy(std::vector<int> y,
   double l = 0.0;
   for (int i = 0; i < y.size(); i++)
   {
-    l += (y[i] * qi[i]);
+    l += (y[i] * log10(qi[i]));
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp]: Use natural log here

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Add docs for Gaussian NB and KNN also

File: docs/methods/linear_model/logistic_regression.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# Logistic Regression
+
+Logistic Regression model using stochastic gradient descent.
```
Comment (uttammittal02) [inline docs/methods/linear_model/logistic_regression.md:3]: It's mini-batch gradient descent and not stochastic.

File: docs/methods/linear_model/logistic_regression.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+LogisticRegression fits a linear model with coefficients $w\  =\  [(w_1, w_2, … , w_p),(w_1, w_2, … , w_p)..]$, where $p$ is number of features and height will be equals to number of classes, to minimize the residual sum of squares between the observed targets in the dataset, and the probabilities predicted by the logistic approximation.
+
+## Parameters
+
+| Name          | Definition                                                                                  | Defaults | Type     |
+| ------------- | ------------------------------------------------------------------------------------------- | -------- | ---------|
+| Epochs        | Number of times that the learning algorithm will work through the entire training dataset   | 100      | `int`    |
+| alpha         | Determines the step size at each iteration while trying to minimize the loss function       | 0.01     | `double` |
+| Batch size    |  Number of x values in batch over which loss will be computed and weights will be updated   | 5        | `int`    |
```
Comment (uttammittal02) [inline docs/methods/linear_model/logistic_regression.md:13]: Change the default batch size to 10% of the whole dataset

File: docs/methods/linear_model/logistic_regression.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+| Name         | Definition                                                | Shape      |
+| ------------ | --------------------------------------------------------- | ---------- |
+| beta | weights of model  | n_features*output_classes |
+
+## Methods
+
+| Name                            | Definition                                            | Return value      |
+| ------------------------------- | ----------------------------------------------------- | ----------------- |
+| `softmax(vector<T> x)`          | Computes softmax                                      | `vector<T>`       |
+| `crossEntropy(vector<int>y,vector<T>qi)` | actual*predicted value                       | DOUBLE       |
```
Comment (uttammittal02) [inline docs/methods/linear_model/logistic_regression.md:26]: `actual * log(predicted)` and not `actual * predicted`

File: docs/methods/linear_model/logistic_regression.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+| `EvalL(vector<vector<T>>x,vector<vector<int>>y,vector<vector<T>>beta)`| Computes loss  |      DOUBLE       |
+| `logRegSgd(vector<vector<double>>x,vector<vector<int>>y,double alpha,int numEpochs,bool verbose)`| Performs logistic regression  |      `vector<vector<T>>`      |
```
Comment (uttammittal02) [inline docs/methods/linear_model/logistic_regression.md]: Change the name for this function also accordingly

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp (hunk 5 above)
Comment (uttammittal02) [reply src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:207]: You need not write your own one-hot encoder. Use the one already written in our library. Also change the name of `train` function to `fit`

File: docs/methods/linear_model/logistic_regression.md (hunk 10 above)
Comment (uttammittal02) [reply docs/methods/linear_model/logistic_regression.md]: Also, add the batch_size parameter here and also in the example below

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    {
+      prm[j] = j;
+    }
+
+    std::random_shuffle(prm.begin(),
+                        prm.end()); // random shuffling of prm vector
+
+    int countCompleted = 0;
+
+    std::vector<std::vector<double>> gradLi( // Initializing gradient descent
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:121]: Declare this vector inside the while loop itself to avoid emptying it each time

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        numClasses, std::vector<double>(x[0].size(), 0.0));
+
+    while (!prm.empty()) // till prm not becomes empty keep iterating over the
+                         // while loop
+    {
+      if ((prm.size() + countCompleted) <
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:127]: This isn't how it actually works.
In each epoch, you need to choose a random value (the starting point of that batch) and take a batch of the given size from that....many training inputs might be considered multiple times in such a case.

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+// This function performs logistic regression stochastic gradient descent
+std::vector<std::vector<double>>
+LogisticRegression<T>::logRegSgd(std::vector<std::vector<double>> x,
+                                 std::vector<std::vector<int>> y, double alpha,
+                                 int numEpochs, bool verbose, int batchSize)
+{                               // PERFORMS MULTICLASS LOGISTIC REGRESSION
+  int n = x.size();             // Rows in training vector
+  int d = x[0].size();          // Number of features
+  int numClasses = y[0].size(); // Output classes
+  std::vector<std::vector<double>> beta(
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:87]: Initialise betas with random values instead of zeros

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.hpp [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+   * @brief computes loss
+   *
+   * @param x training values
+   * @param beta weights
+   * @return double loss value
+   */
+  double EvalL(std::vector<std::vector<T>> x, std::vector<std::vector<int>> y,
+               std::vector<std::vector<T>> beta);
+  /**
+   * @brief Logistic regression
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.hpp:42]: Change it to `Mini-batch gradient descent`.
Also, add the batch-size parameter here

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
A few more changes

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+// x -> training 2-d vector double values
+// y - > training 1-d int values
+// alpha -> double which is going to be multiply with gradient descent
+// numEpochs -> number of epochs
+// verbose -> whether to show which epoch is going on
+void LogisticRegression<T>::fit(std::vector<std::vector<T>> x,
+                                std::vector<int> y, double alpha, int numEpochs,
+                                bool verbose, int batchSize)
+{
+  this->batchSize = batchSize;
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.cpp:226]: where did you set the default value for the batch_size

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Some old changes are left and one or two new changes have been requested

File: docs/methods/linear_model/logistic_regression.md (hunk 8 above)
Comment (uttammittal02) [reply docs/methods/linear_model/logistic_regression.md:13]: @Sar-thak-3 do it bro

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.hpp [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+  public:
+  /**
+   * @brief training
+   *
+   * @param x training x values
+   * @param y training y values
+   * @param alpha double value which multiplied with gradient descent
+   * @param numEpochs number of epochs
+   * @param batchSize size of match
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.hpp:66]: match?

File: src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.hpp [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+   * @param verbose require to print which epoch is going on
+   */
+  void fit(std::vector<std::vector<T>> x, std::vector<int> y, double alpha,
+           int numEpochs, bool verbose, int batchSize = -1);
+  /**
+   * @brief Predict value for testing
+   *
+   * @param x testing value
+   * @return std::vector<double> one hot encoded probobabilities for each class
```
Comment (uttammittal02) [inline src/slowmokit/methods/linear_model/logistic_regression/logistic_regression.hpp:75]: fix the typo in `probobabilities`

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Just some last changes

**Review by uttammittal02 — APPROVED** [review]
Lgtm....op work
